concrete-core/src/backends/core/private/math/decomposition/iter.rs

```diff
@@ -313,17 +313,30 @@ fn decompose_one_level<S: UnsignedInteger>(
 ) -> (S, S) {
     // We perform the division of the input by q/B^i
     let res = input >> (S::BITS - base_log * current_level);
+    // And by q/B^(i+1)
+    let next_res = res >> base_log;
     // We reduce the result modulo B
     let res = res & mod_b_mask;
     // The result may already be greater or equal to B/2.
     let carry = res & carry_mask;
+    // We check if the next result will be greater or equal to B/2
+    let next_carry = next_res & carry_mask;
     // We propagate the carry from the previous level
     let res = res.wrapping_add(previous_carry);
     // The previous carry may have made the result larger or equal to B/2.
     let carry = carry | (res & carry_mask);
-    // If the result is greater or equal to B/2, we subtract B from the result (viewed as a
-    // signed integer)
-    let res = (res.into_signed() - (carry << 1).into_signed()).into_unsigned();
```

Comment:
I took advantage of this PR to replace the circonvoluted line above by a more readable wrapping_sub (line 339). I supposed this does the exact same thing, am I wrong?